### DISCUSSION TITLE
Github Actions - git command `ls-remote` fails for private repo when on runner (works locally)

### DISCUSSION THREAD

**Discussion by aormsby** [community/community#25240] [Actions]
<p>I’m using this command to check and see if a remote <strong>private</strong> repo exists and the user has access to it:</p>
<p><code>git ls-remote https://&lt;user name&gt;:&lt;token&gt;@github.com/&lt;repo owner&gt;/&lt;repo name&gt;.git</code></p>
<p>Also checking for a branch on the remote repo:</p>
<p><code>git ls-remote &lt;same url&gt; &lt;branch name&gt;</code></p>
<p>These <strong>should</strong> be able to access the private repo using the access token passed into the action, and I should get an actual read on the remote repo and its refs. However, I receive a git error:</p>
<p><code>remote: Repository not found. fatal: repository 'https://github.com/aoAppDev/upstream-sync-test-remote.git/' not found</code></p>
<p><strong>Notably</strong>, this output doesn’t include the classic *** to hide the token in the output URL because, well… it isn’t in this error URL at all.</p>
<p>It’s not an issue with the token or URL because the action successfully clones the private repo <strong>immediately after</strong> these commands fail. (Also, it works as expected on my local machine.)</p>
<p><strong>So why would this command fail on the runner?<br>
Is there some config I’m missing?<br>
Something blocking the use of ls-remote with a secret token??</strong></p>
<p>I’m at a loss here.</p>

**Comment by aormsby** [comment, marked as answer]
<p>I figured out the problem! I was using the GH checkout action as a first step, and the GITHUB_TOKEN credentials were automatically persisting through the entire workflow. So anytime I tried to run a command requiring the auth token for the second remote repo, it was using the persisted credentials. I set <code>persist-credentials</code> to false, and my commands work as expected.</p>
<p>Some odd things about this, though –</p>
<ul>
<li>If I have the auth token in the remote URL, I don’t see why the action would even try to use the persisted auth token.</li>
<li>Why did the clone work while the other commands didn’t? Super strange.</li>
</ul>
<p>Anyway, I hope this helps someone down the line. Happy coding!</p>

**Reply by dodomood** [reply]
Spent almost 8h today on this shit. Thank you for existing! 🥇

**Reply by pedroscaff** [reply]
I also spent a few hours on this, unbelievable that this is the default behavior :cry:. Thank you very much!

**Reply by robindowling** [reply]
I spent hours on this as well, thank you for sharing and allowing me to finally move on from this 😊

**Reply by dan-developer** [reply]
Thank you for existing! 🥇

**Reply by sugarlata-zetifi** [reply]
Champion - thanks for the fix! Way too much time on this

**Reply by jebarrocal** [reply]
I spend too many hours with this shit.
Love you. Thank you for share this.

**Reply by galnadjar** [reply]
spent many hours on this crap, thx !

**Reply by FJCC** [reply]
This answer has made two happy on the same day! Thanks!

**Reply by Bok-GiUng** [reply]
I can't speak English well because I'm Korean, but I want to thank you. The hours with GPT5 were terrible. You saved my life.

**Reply by noamnelke** [reply]
Like many before me on this thread, this brought 4 gruelling hours with ChatGPT to an end. Thanks for taking the time to share your findings.

FWIW, I just used this in the checkout step, so all interactions use the token:
```
      - name: Checkout
        uses: actions/checkout@v4
        with:
          token: ${{ secrets.GH_TOKEN }}
```

**Comment by nikitaeverywhere** [comment]
<p>Thanks for sharing this! Saved a few hours of debugging for sure.</p>

**Comment by Nokel81** [comment]
Thank you sooo much!!! I have the same set up it seems where I call `git ls-remote` (with the auth in the URL) which would fail for private repos. But if I tried to clone a specific tag it would work.

**Comment by bloodearnest** [comment]
Thanks for this. We had accidentally included the .git/ dir in our docker image, and then git ls-remote was failing in production due to the cached credentials.

**Comment by AmitBenAmi** [comment]
Had this issue today as well, and I've been running within a Dockerfile and was passing the `GITHUB_TOKEN` as a secret to the action as an input - so I thought it doesn't even related - I'm not running the commands directly within the workflow.

Turning the persistency off solved my issue, so thanks!
